File: ruby/81_search_in_rotated_sorted_array_two.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Boolean}

# Input: nums = [2,5,6,0,0,1,2], target = 0
# Output: true

# 给定一个旋转过的升序数组，要求寻找target，返回bool，此时数组可能有重复元素
# 思路同 33题

# 本题使用的是与right相比，其实道理和用left是一样的
def search(nums, target)
    left, right = 0, nums.size-1
    while left <= right
    	mid = left + (right-left)/2

    	return true if nums[mid] == target

    	if nums[mid] < nums[right]
    		if target > nums[mid] && target <= nums[right]
    			left = mid+1
    		else
    			right = mid-1
    		end
    	elsif nums[mid] > nums[right]
    		if target < nums[mid] && target >= nums[left]
    			right = mid-1
    		else
    			left = mid+1
    		end
    	else
    		right -= 1
    	end
    end

    false
end


# 另一种思路，因为数组被旋转过，所以找到数组的首元素，这样数组也拆分为两个有序的数组，此时在按照原始的二分查找即可

def find_nums_head(nums)
	for i in 1..nums.size-1
		return i if nums[i] < nums[i-1]
	end

	return 0
end

def search_version2(nums, target)
	head = find_nums_head(nums)

	if head == 0
		return nums.bsearch { |i| target <=> i } != nil
	else
		return nums.slice(0...head).bsearch { |i| target <=> i  } != nil ||
		       nums.slice(head...nums.size).bsearch { |i| target <=> i } != nil
	end
end


nums = [2,5,6,0,0,1,2]
puts search_version2(nums, 0)

File: cpp/81_search_in_rotated_sorted_array_two.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 同33，给定rotated sorted array(可能存在重复元素)和target，要求查找是否存在；
class Solution {
public:
    bool search(vector<int>& nums, int target) {
		if (nums.empty()) { return false; }
		int left = 0, right = nums.size()-1;

		while (left <= right) {
			int mid = left + (right - left) / 2;
			if (nums[mid] == target) { return true; }

			if (nums[left] < nums[mid]) { // left sorted
				if (target < nums[mid] && nums[left] <= target) {
					right = mid - 1;
				} else {
					left = mid + 1;
				}
			} else if (nums[left] > nums[mid]) {
				if (target <= nums[right] && nums[mid] < target) {
					left = mid + 1;
				} else {
					right = mid - 1;
				}
			} else {
				++left;
			}
		}

		return false;
    }
};
